JDK-8301661: Enhance os::pd_print_cpu_info on macOS and Windows

MBaesken opened the pull request:
Enhance os::pd_print_cpu_info on macOS and Windows by information about CPU frequency and caches.
On Windows , this info can be obtained by the Processor Power Information API or "powerbase" (CallNtPowerInformation , see https://learn.microsoft.com/en-us/windows/win32/api/powerbase/nf-powerbase-callntpowerinformation ); this is available since Windows Server 2003/XP.
On macOS, sysctlbyname can be used.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8301661](https://bugs.openjdk.org/browse/JDK-8301661): Enhance os::pd_print_cpu_info on macOS and Windows


### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [9a311018](https://git.openjdk.org/jdk/pull/12403/files/9a311018c1873ae6f63ff158e1db4d4284bcb959)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**) ⚠️ Review applies to [9a311018](https://git.openjdk.org/jdk/pull/12403/files/9a311018c1873ae6f63ff158e1db4d4284bcb959)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12403/head:pull/12403` \
`$ git checkout pull/12403`

Update a local copy of the PR: \
`$ git checkout pull/12403` \
`$ git pull https://git.openjdk.org/jdk pull/12403/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12403`

View PR using the GUI difftool: \
`$ git pr show -t 12403`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12403.diff">https://git.openjdk.org/jdk/pull/12403.diff</a>

</details>

dholmes-ora commented:
Okay the usual question when we augment the information that will appear in hs_err file: how safe is this going to be to call from a signal handler?

MBaesken commented:
> Okay the usual question when we augment the information that will appear in hs_err file: how safe is this going to be to call from a signal handler?

Hi David , I think the sysctl / sysctlbyname calls on macOS are rather safe (and we use this already for some time for other entries e.g. in os::print_summary_info  in vm error reporting, so nothing really new. Not so sure about the Windows powerprof / powerbase API, this was not used before to my knowledge in OpenJDK. However we use both macOS and Windows APIs for some time in our internal JVM, no issues seen so far.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1876,8 +1877,40 @@ void os::win32::print_windows_version(outputStream* st) {
+
 void os::pd_print_cpu_info(outputStream* st, char* buf, size_t buflen) {
-  // Nothing to do for now.
+  SYSTEM_INFO si;
+  GetSystemInfo(&si);
+
+  PROCESSOR_POWER_INFORMATION ppi;
+  size_t sz_check = sizeof(ppi) * (size_t)si.dwNumberOfProcessors;
```
That said from what I'm reading if systems have more than 64 processors they will be divided into processor groups and the max number will be 64. However IIUC `CallNtPowerInformation` will expect a buffer big enough for the true number of logical processors >64.

MBaesken commented:
Hi David, thanks for the comments. I did a few adjustments following your comments. Regarding 

> Not sure about this mechanism versus adding it to Makefile ??

Both works. From what I see, currently we use the Makefile to set the lib deps, should I remove the pragma and use the Makefile ?
Regarding the processor count , probably we do not want to malloc anyway in  pd_print_cpu_info so we have to live with the space  buf  provides.
Btw what you think about shortening the info, e.g. check if all processors have the same (M)Hz values and if it is the case just print once, this might look better and much less redundant in the output ?

dholmes-ora commented:
> From what I see, currently we use the Makefile to set the lib deps, should I remove the pragma and use the Makefile ?

I'm asking the build folk if they have any preference/concerns.

> Regarding the processor count , probably we do not want to malloc anyway in pd_print_cpu_info so we have to live with the space buf provides.

True. I still think we might reuse processor_count(). Showing zero very early during VM init is not really an issue.

> Btw what you think about shortening the info, e.g. check if all processors have the same (M)Hz values and if it is the case just print once, this might look better and much less redundant in the output ?

As this is intended as fairly general information shortening would be good - especially if the values reported are the same. Do we need to report all processors, or can we just use one as representative?

MBaesken commented:
Hi David , I did another small adjustment. Reused  os::processor_count if it delivers a value larger than 0. And adjusted the output a bit. 

>  especially if the values reported are the same

On my notebook I noticed that the current idle state sometimes differs among processors. So it is not always the same.
Maybe it makes sense to print the current idle state separately for all processors and  if possible,  output the other values just once .

`
ProcessorInformation for all processors:
  Max Mhz: 2611, Current Mhz: 2611, Mhz Limit: 2611
  MaxIdleState: 2
CurrentIdleState for processors 0 - 15: 2 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2
`

dholmes-ora commented:
With regards to the pragma I had the following response from @magicus :
> Oh, please don't do that. We could perhaps have gone down that route, but we didn't. If you do that, there won't be a single source of truth on where to look for library linkage. We had a bug a year or two ago that ended up in us loading a library which nobody expected or realized due to a windows pragma. So, basically, there's nothing inherently wrong with this (just a bit non-standard from an old-school unix point of view), but it does not fit with the current model of the JDK

dholmes-ora commented:
> On my notebook I noticed that the current idle state sometimes differs among processors. So it is not always the same.
Maybe it makes sense to print the current idle state separately for all processors and if possible, output the other values just once .

I have no idea what possible use the "idle state" information  would be in any case. We don't have that per-processor info on non-Windows.

magicus commented:
/label build

magicus reviewed (approved):
Build changes look good now. Thanks for removing the pragma.

/reviewers 2

MBaesken commented:
> I have no idea what possible use the "idle state" information would be in any case. We don't have that per-processor info on non-Windows.

Hi David,  there is some info about the idle states at  https://learn.microsoft.com/en-us/windows-hardware/test/wpt/cpu-analysis  ;  if you think that currently, with the info only available on Windows, it makes not that much sense to add, it I can omit it for now.  Would probably make it easier to scan through the cpu-infos and observe same MHz numbers for all and then print the same info for all processors (as I said, where I tested the MHz values where the same for all CPUs only the current idle states varied).

MBaesken commented:
Hi Magnus, thanks for the review !

RealLucy reviewed (approved):
Changes look good to me

MBaesken commented:
Hi David are you okay with the latest version ?

dholmes-ora commented:
> Hi David are you okay with the latest version ?

I thought you were looking at reducing the output:

> Maybe it makes sense to print the current idle state separately for all processors and if possible, output the other values just once .

I don't think idle states are particularly useful, so I'd prefer to see similar output to non-Windows rather than per-processor info that is the same for everything of interest.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1876,8 +1877,64 @@ void os::win32::print_windows_version(outputStream* st) {
+    }
+
+    if (same_vals_for_all_cpus && MaxMhz != -1) {
+      st->print_cr("ProcessorInformation for all %d processors :", proc_count);
```
Nit: Processor Information

MBaesken replied:
I adjusted the output.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1876,8 +1877,64 @@ void os::win32::print_windows_version(outputStream* st) {
+        if (MaxMhz != (int) pppi->MaxMhz ||
+            CurrentMhz != (int) pppi->CurrentMhz ||
+            MhzLimit != (int) pppi->MhzLimit) {
+          same_vals_for_all_cpus = false;
```
You could break once this is false.

MBaesken replied:
Added a break.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1876,8 +1877,64 @@ void os::win32::print_windows_version(outputStream* st) {
+  }
+
+  size_t sz_check = sizeof(PROCESSOR_POWER_INFORMATION) * (size_t)proc_count;
+  NTSTATUS status = ::CallNtPowerInformation(ProcessorInformation, NULL, 0, buf, (ULONG) buflen);
```
How is the caller supposed to know what size buffer to pass? Are they just supposed to guess and hope it is big enough??

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1876,8 +1877,64 @@ void os::win32::print_windows_version(outputStream* st) {
+
+  size_t sz_check = sizeof(PROCESSOR_POWER_INFORMATION) * (size_t)proc_count;
+  NTSTATUS status = ::CallNtPowerInformation(ProcessorInformation, NULL, 0, buf, (ULONG) buflen);
+  int MaxMhz = -1, CurrentMhz = -1, MhzLimit = -1;
```
Nit: variable names should not start with Capital letters

MBaesken replied:
I adjusted the variable names.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1876,8 +1877,64 @@ void os::win32::print_windows_version(outputStream* st) {
+        }
+      }
+      // avoid iteration in case buf is too small to hold all proc infos
+      if (sz_check > buflen) break;
```
In this case shouldn't the function have returned `STATUS_BUFFER_TOO_SMALL`?

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1876,8 +1877,64 @@ void os::win32::print_windows_version(outputStream* st) {
+    // differing values, iterate again
+    pppi = (PROCESSOR_POWER_INFORMATION*) buf;
+    for (int i = 0; i < proc_count; i++) {
+      st->print_cr("ProcessorInformation for processor %d", (int) pppi->Number);
```
Nit: Processor Information

dholmes-ora reviewed (commented):
General approach seems okay. A couple of queries and nits.

Thanks.

MBaesken commented:
Hi David I adjusted the output, renamed some variables and added a break where you suggested.

> In this case shouldn't the function have returned STATUS_BUFFER_TOO_SMALL

I think this would be the case, but I was not sure so I added the check.

> How is the caller supposed to know what size buffer to pass? Are they just supposed to guess and hope it is big enough??

I think this is the same for most of the print_  functions from
src/hotspot/share/runtime/os.hpp
getting a buffer 'buf'  and a buf-length value, they are unfortunately not very specific on the size passed.

https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/os.hpp#L754

  // Print out system information; they are called by fatal error handler.
  // Output format may be different on different platforms.

dholmes-ora reviewed (approved):
I'm still a little dubious about the buffer size check but lets see how this goes.

Thanks.

MBaesken commented:
Hi David and Lutz, thanks for the reviews.

/integrate
